File: emulator/controller_emul/pump_protocol.py
# -*- coding:utf-8 -*-
import os
import sys
import signal
import getopt
import time
import struct
import logging
import copy
import socket
import struct
import base64
import http.client
import urllib.parse
import controller_emul.config
import controller_emul.enum


class PumpMessageBase:
    def __init__ (self, type):
        self.type = type

    def serialize(self):
        fmt = "B"
        return  struct.pack(fmt, self.type)



class PumpMessageCommandWithId (PumpMessageBase):
    def __init__ (self, type, id):
        super().__init__(type)
        self.id = id

    def serialize(self):
        fmt = "!BI"
        return struct.pack(fmt, self.type, self.id)



class ControllerInfoCommand(PumpMessageCommandWithId):
    def __init__ (self, type, id, token):
        super().__init__(type, id)
        self.token = token

    def serialize(self):
        fmt = "!BI"
        res = struct.pack(fmt, self.type, self.id) + self.token
        return res


class CheckCommandRequest(PumpMessageBase):
    def __init__(self, type, token):
        super().__init__(type)
        self.token = token

    def serialize(self):
        fmt = "B"
        res = struct.pack(fmt, self.type) + self.token
        return res
 #
 #    uint32_t    pressure;    //current pressure
 #    bool is_working;         //is controller working
 #    uint8_t currentValve;    //current valve opened
 #    uint8_t currentStep;     //current step

class CheckCommandWithInfoRequest(PumpMessageBase):
    def __init__(self, type, token, pressure, is_working, current_valve, current_step):
        super().__init__(type)
        self.token = token

        self.pressure = pressure
        self.is_working = is_working
        self.current_valve = current_valve
        self.current_step = current_step

    def serialize(self):
        fmt = "!BIBBB"
        return struct.pack(fmt, self.type, self.pressure, self.is_working, self.current_valve, self.current_step) + self.token

class PumpProtocol (object) :

    MESSAGE_TYPE = controller_emul.enum.enum(
        PUMP_COMMAND_CHECK = 0x30,
        NO_COMMAND_RESPONSE = 0x31,   #nothing to do
        GET_INFO_RESPONSE = 0x32,     #controller must return information
        SEND_INFO_REQUEST = 0x33,     #send info
        SWITCH_TO_MONITORING_MODE_RESPONSE = 0x34,
        COMMAND_CHECK_WITH_INFO_REQUEST=0x35

    )

    def __init__(self, url) :
        self.urlParsed = urllib.parse.urlsplit(url)

    def parse_response(self,response):
        length = len(response)
        offset = 0
        type = None
        if length >= offset + 1:
            type = response[offset]

        res = None
        offset += 1

        if type == self.MESSAGE_TYPE.NO_COMMAND_RESPONSE:
            res = PumpMessageBase(type)
        elif type == self.MESSAGE_TYPE.GET_INFO_RESPONSE:
            fmt = "!I"
            data = response[offset:]
            if struct.calcsize(fmt) == len(data):
                res = PumpMessageCommandWithId(type, struct.unpack(fmt, data)[0])

        elif type == self.MESSAGE_TYPE.SWITCH_TO_MONITORING_MODE_RESPONSE :
            res = PumpMessageBase(type)

        return res


    def send_command_request(self, message):


        path = self.urlParsed.path + "?" + base64.b64encode(message).decode('latin-1')

        res = None

        try:

            conn = http.client.HTTPConnection(self.urlParsed.netloc)
            conn.request("GET", path, None, {})

            response = conn.getresponse()

            if response.status != 204:
                data = response.read()
            controller_emul.LOGGER.debug("Send command %s,  return code %d  response %s", path, response.status, data)
            res = self.parse_response(data)
        finally:
            conn.close()

        return res

    def send_controller_info(self, commandId, token, info):
        command = ControllerInfoCommand(self.MESSAGE_TYPE.SEND_INFO_REQUEST, commandId, token )

        return self.send_command_request(command.serialize())



    def send_check_command_request(self, token):
        command = CheckCommandRequest(self.MESSAGE_TYPE.PUMP_COMMAND_CHECK, token)

        return self.send_command_request(command.serialize())

File: web/framework/pumpprotocol.php
<?php
/**
 * Created by PhpStorm.
 * User: user
 * Date: 26.09.18
 * Time: 17:41
 */
require_once("monitoringinfo.php");
abstract class PumpMessageConsts
{
    public static $COMMAND_CHECK_REQUEST = 0x30;
    public static $NO_COMMAND_RESPONSE = 0x31; //nothing to do
    public static $GET_INFO_RESPONSE = 0x32;   //controller must return information
    public static $SEND_INFO_REQUEST = 0x33;   //controller return information about self
    public static $SWITCH_TO_MONITORING_MODE_RESPONSE = 0x34; //controller must switch for full monitoring mode response
    public static $COMMAND_CHECK_WITH_INFO_REQUEST = 0x35;   //check command with full monitoring info


}




class PumpMessageBase
{
    protected $type;

    public function __construct($type)
    {
        $this->type = $type;
    }

    public function getType()
    {
        return $this->type;
    }

    public function serialize()
    {
        return chr($this->type);

    }


}

;

/**
 *  Basic request from controlServer
 *  Format:
 *  struct BasicRequest
 *  {
 *    uint8_t messageType; //one of PumpMessageCodes
 *    char token[];        //tokenSize
 * }
 **/
class BasicRequest extends PumpMessageBase
{
    protected static $MAX_TOKEN_SIZE = 20;
    protected $token;

    public function __construct($type, $token)
    {
        parent::__construct($type);
        $this->token = $token;


    }

    public function getToken()
    {
        return $this->token;

    }

    static public function deserialize($type, $data)
    {
        if (is_string($data) && strlen($data) <= BasicRequest::$MAX_TOKEN_SIZE)
            return new BasicRequest($type, $data);
        else
            return null;

    }

}

/**
 * Class SendMonitoringInfoRequest
 * Returns breef monitoring info from controller
 * Format:
 *  struct SendMonitoringInfoRequest
 *  {
 *    uint8_t     messageType; //one of PumpMessageCodes::$SWITCH_TO_MONITORING_MODE_RESPONSE
 *    char token[];        //tokenSize
 * }
 *
 */

class SendInfoRequest extends BasicRequest
{

    private $commandId;
    public $pumpingTable = NULL;

    public function __construct($type, $token, $commandId)
    {
        parent::__construct($type, $token);
        $this->commandId = $commandId;
    }

    public function getCommandId()
    {
        return $this->commandId;

    }

    static public function deserialize($type, $data)
    {
        if (!is_string($data))
            return null;


        $offset = 0;
        $len = strlen($data);
        $commandId = null;
        $token = null;

        $mailformed = false;
        if ($len > $offset + 4) {
            $commandId = unpack("N", substr($data, $offset, 4))[1];
        } else {
            $mailformed = true;
        };

        $offset += 4;

        if (!$mailformed && $len >= $offset + 1 && ($len - $offset) < BasicRequest::$MAX_TOKEN_SIZE) {
            $token = substr($data, $offset, $len - $offset);

        } else {

            $mailformed = true;
        };

        $res = null;
        if (!$mailformed) {
            $res = new  SendInfoRequest($type, $token, $commandId);

        };

        return $res;

    }

}

/**
 * Class SendMonitoringInfoRequest
 * Returns breef monitoring info from controller
 * Format:
 *  struct SendMonitoringInfoRequest
 *  {
 *    uint8_t     messageType; //one of PumpMessageCodes::$SWITCH_TO_MONITORING_MODE_RESPONSE
 *    uint32_t    pressure;    //current pressure
 *    bool is_working;         //is controller working
 *    uint8_t currentValve;    //current valve opened
 *    uint8_t currentStep;     //current step
 *    char token[];        //tokenSize
 * }
 *
 */
class SendMonitoringInfoRequest extends BasicRequest
{
    public $monitoringInfo = null;

    public function __construct($type, $monitoringInfo, $token)
    {
        parent::__construct($type, $token);
        $this->monitoringInfo = $monitoringInfo;
    }


    static public function deserialize($type, $data)
    {
        if (!is_string($data))
            return null;


        $offset = 0;
        $len = strlen($data);
        $monitoring_info = null;
        $token = null;

        $mailformed = false;
        if ($len > $offset + MonitoringInfo::getStructFMTSize()) {
            $monitoring_info = MonitoringInfo::deserialize( substr($data, $offset, MonitoringInfo::getStructFMTSize() ));
        } else {
            $mailformed = true;
        };

        $offset += MonitoringInfo::getStructFMTSize();

        if (!$mailformed && $len >= $offset + 1 && ($len - $offset) < BasicRequest::$MAX_TOKEN_SIZE) {
            $token = substr($data, $offset, $len - $offset);

        } else {

            $mailformed = true;
        };

        $res = null;
        if (!$mailformed) {
            $res = new  SendMonitoringInfoRequest($type, $monitoring_info, $token);

        };

        return $res;


    }
}


class BasicResponse extends PumpMessageBase
{

    public function __construct($type)
    {
        parent::__construct($type);
    }


}

/**
 * Controller must to the command
 */
class PendingCommandResponse extends BasicResponse
{
    public $command_id;

    public function __construct($type, $command_id)
    {
        parent::__construct($type);
        $this->command_id = $command_id;

    }

    public function serialize()
    {
        return pack("CN", $this->getType(), $this->command_id);
    }
}


function pumpProtocolMessageFromBytes($data)
{
    $res = null;
    $messageType = null;
    $clientToken = null;
    $len = strlen($data);
    $mailfomed = false;
    $tokenSize = 0;
    $dataSize = 0;
    $offset = 0;

    //extract messageType
    $t = is_string($data);
    if (is_string($data) && $len >= $offset + 1) {
        $messageType = ord($data[0]);
    } else
        $mailfomed = true;

    $offset += 1;
    if (!$mailfomed) {
        switch ($messageType) {
            case  PumpMessageConsts::$COMMAND_CHECK_REQUEST :
                $res = BasicRequest::deserialize($messageType, substr($data, $offset, $len - $offset));
                break;

            case PumpMessageConsts::$SEND_INFO_REQUEST:
                $res = SendInfoRequest::deserialize($messageType, substr($data, $offset, $len - $offset));
                break;

            case PumpMessageConsts::$COMMAND_CHECK_WITH_INFO_REQUEST:
                $res = SendMonitoringInfoRequest::deserialize($messageType, substr($data, $offset, $len - $offset));
                break;


        };


    };
    return $res;


};

?>